### PR TITLE
Feature vram opt

### PR DESCRIPTION
**Pull request by MartinMao2023**
Related issues: PGA and DCRL: duplicated sampling of transitions in "emit_pg", unnecessary occupation of Vram in "state_update", too complicated structure of functions.

This PR introduces:
- Off-springs performing PG mutation now share the same mini-batch of data in emit_pg. 
- Sampling of transitions in both "emit_pg" and "state_update" is now made step by step. 
- "policy_loss" in "make_td3_dc_loss" now take desc_prime from the input instead of from transitions (so that transition can be shared) .
- Function "mutate_function_pg" is simplified and emerged into "emit_pg".
- Combined "train_xxxx" and "update_xxxx" into a single function.
- Build a new function "scan_actor_critic_training" for Actor-Critic training loop.
- Removed meaningless similarity (and novelty) calculation in dcrl_emitter (and in dpg_emitter).
- dpg_emitter modified accordingly.

## Checks
 
- [x] a clear description of the PR has been added
- [x] sufficient tests have been written
- [x] example notebook have been added to the repo
- [x] all example notebooks have been verified to execute without errors
- [x] clean docstrings and comments have been written
- [ ] relevant section have been added to the documentation
- [x] if any issue/observation has been discovered, a new issue has been opened
- [x] the source branch is `develop` if this PR introduces a new feature

## Benchmark Test Results

All updated emitters have been tested with corresponding example Jupyter notebooks. No error or loss of performance observed.

DCRL (Ant_omni,  128 GA + 64 PG + 64 AI)
New version (11min16s):
new_dcrl_ant_omni_(11min30s)
Old version (11min23s):
old_dcrl_ant_omni_(11min23s)

DCRL (Walker2d, 128 GA + 64 PG + 64 AI)
New version (12min32s):
new_dcrl_walker_(12min32s)
Old version (12mins45s):
old_dcrl_walker_(12min45s)

PGA (Ant_omni, 64 GA + 63 PG + 1 AI)
New version (12min44s):
new_pga_ant_omni_(12min44s)
Old version (13min07s):
old_pga_ant_omni_(13min07s)

PGA (Walker2d, 64 GA + 63 PG + 1 AI)
New version (14min06s):
new_pga_walker_(14min6s)
Old version (14min31s):
old_pga_walker_(14min31s)

QDPG is super slow (5 seconds each iteration), main bottle-neck is the diversity archive insertion. Benchmark excluded.

All tests done with 1000 iterations, minibatch size 256, 150 PG mutation steps, 3000 critic training steps, 2 for policy delay, and seed 42. Ant_omni takes 250 time-steps and walker takes 1000.

## Future improvements
 
DPG and QPG in QDPG use two but actually duplicated replay-buffers, which can be merged into a single one to save several GiBs of Vram. Advise optimize runtime performance for diversity archive insertion, which is taking 80% of the time.

### CODE REVIEW

File: qdax/core/emitters/dcrl_emitter.py [hunk 1]
```diff
@@ -293,33 +290,34 @@ def emit(
...
-        descs_pg = selected_sub_repertoire.descriptors
+        parents_pg = sub_repertoire.genotypes
+        descs_pg = sub_repertoire.descriptors
         genotypes_pg = self.emit_pg(emitter_state, parents_pg, descs_pg)
 
         # Actor injection emitter
-        key, subkey = jax.random.split(key)
         descs_ai = repertoire.select(
-            key, self._config.ai_batch_size, selector=self._selector
+            subkey, self._config.ai_batch_size, selector=self._selector
```
Comment (Lookatator) [inline qdax/core/emitters/dcrl_emitter.py:302]: I'd suggest to do a 
```python
key, subkey = jax.random.split(key)
```
every time, and only pass the subkeys (never the key) to the called functions. Just to remain consistent with the rest of the code

File: qdax/core/emitters/qpg_emitter.py [hunk 2]
```diff
@@ -125,7 +124,7 @@ def use_all_data(self) -> bool:
 
     def init(
         self,
-        key: RNGKey,
+        random_key: RNGKey,
```
Comment (Lookatator) [inline qdax/core/emitters/qpg_emitter.py:127]: the name of random keys is usually `key`

File: qdax/core/emitters/qpg_emitter.py [hunk 3]
```diff
@@ -136,27 +135,27 @@ def init(
 
         Args:
             genotypes: The initial population.
-            key: A random key.
+            random_key: A random key.
 
         Returns:
-            The initial state of the PGAMEEmitter.
+            The initial state of the PGAMEEmitter, a new random key.
```
Comment (Lookatator) [inline qdax/core/emitters/qpg_emitter.py:141]: the function does not return a random key

File: qdax/core/emitters/qpg_emitter.py [hunk 4]
```diff
@@ -136,27 +135,27 @@ def init(
...
         # Initialise critic, greedy actor and population
-        key, subkey = jax.random.split(key)
+        random_key, subkey = jax.random.split(random_key)
         fake_obs = jnp.zeros(shape=(observation_size,))
         fake_action = jnp.zeros(shape=(action_size,))
         critic_params = self._critic_network.init(
             subkey, obs=fake_obs, actions=fake_action
         )
-        target_critic_params = jax.tree.map(lambda x: x, critic_params)
+        target_critic_params = jax.tree_util.tree_map(lambda x: x, critic_params)
```
Comment (Lookatator) [inline qdax/core/emitters/qpg_emitter.py:155]: the new standard is `jax.tree.map`

File: qdax/core/emitters/qpg_emitter.py [hunk 5]
```diff
@@ -189,16 +188,18 @@ def init(
...
 
-    @partial(jax.jit, static_argnames=("self",))
+    @partial(
+        jax.jit,
+        static_argnames=("self",),
+    )
     def emit(
         self,
-        repertoire: GARepertoire,
+        repertoire: Repertoire,
```
Comment (Lookatator) [inline qdax/core/emitters/qpg_emitter.py:202]: Shouldn't it be GARepertoire here?

File: qdax/core/emitters/qpg_emitter.py [hunk 6]
```diff
@@ -210,7 +211,7 @@ def emit(
             key: a random key
 
         Returns:
-            A batch of offspring, the new emitter state.
+            A batch of offspring, the new emitter state and a new key.
```
Comment (Lookatator) [inline qdax/core/emitters/qpg_emitter.py:214]: it only returns an emitter state

File: qdax/core/emitters/qpg_emitter.py [hunk 7]
```diff
@@ -228,20 +229,22 @@ def emit(
         offspring_actor = self.emit_actor(emitter_state)
 
         # add dimension for concatenation
-        offspring_actor = jax.tree.map(
+        offspring_actor = jax.tree_util.tree_map(
```
Comment (Lookatator) [inline qdax/core/emitters/qpg_emitter.py:232]: same remark as above (`jax.tree.map`)

File: qdax/core/emitters/dcrl_emitter.py [hunk 8]
```diff
@@ -602,159 +587,145 @@ def _update_actor(
...
+        return new_emitter_state, ()
 
     @partial(jax.jit, static_argnames=("self",))
-    def _train_policy(
+    def _update_policy(
         self,
-        emitter_state: DCRLEmitterState,
         policy_params: Params,
         policy_opt_state: optax.OptState,
+        desc_prime: jnp.ndarray,
```
Comment (Lookatator) [inline qdax/core/emitters/dcrl_emitter.py:696]: should be of type `Descriptor` ?

File: qdax/core/neuroevolution/losses/td3_loss.py [hunk 9]
```diff
@@ -125,13 +125,12 @@ def make_td3_loss_dc_fn(
     def _policy_loss_fn(
         policy_params: Params,
         critic_params: Params,
+        desc_prime: jnp.ndarray,
```
Comment (Lookatator) [inline qdax/core/neuroevolution/losses/td3_loss.py:128]: Should be of type Descriptor?

File: qdax/core/neuroevolution/losses/td3_loss.py [hunk 10]
```diff
@@ -157,11 +156,12 @@ def _critic_loss_fn(
         target_actor_params: Params,
         target_critic_params: Params,
         transitions: Transition,
-        key: RNGKey,
+        random_key: RNGKey,
```
Comment (Lookatator) [inline qdax/core/neuroevolution/losses/td3_loss.py:159]: same remark as above: should be renamed to `key`

**Review by Lookatator — CHANGES_REQUESTED** [review]

File: qdax/core/emitters/dcrl_emitter.py [hunk 11]
```diff
@@ -187,7 +182,7 @@ def init(
         transitions = extra_scores["transitions"]
         episode_length = transitions.obs.shape[1]
 
-        desc = jnp.repeat(descriptors[:, jnp.newaxis, :], episode_length, axis=1)
+        desc = jnp.repeat(descriptors[:, None, :], episode_length, axis=1)
```
Comment (Lookatator) [inline qdax/core/emitters/dcrl_emitter.py:185]: what is the problem with `jnp.newaxis`? I personally find this clearer

**Comment by Lookatator** [comment]
Thanks @MartinMao2023 for the PR and for all the plots! 😄 
I've just requested a few changes (those are only renaming changes), all the rest looks good to me!

File: qdax/core/emitters/dcrl_emitter.py [hunk 12]
```diff
@@ -274,14 +268,17 @@ def _compute_equivalent_params_with_desc(
...
 
-    @partial(jax.jit, static_argnames=("self",))
+    @partial(
+        jax.jit,
+        static_argnames=("self",),
+    )
     def emit(
         self,
-        repertoire: MapElitesRepertoire,
+        repertoire: Repertoire,
```
[the remark this answers is not in the document]
Comment (MartinMao2023) [reply qdax/core/emitters/dcrl_emitter.py:277]: I have changed that. But there seems to be an abusive mix of use of "MapElitesRepertoire" and "Repertoire". I suggest having them unified to be consistent.

File: qdax/core/emitters/dcrl_emitter.py (hunk 1 above)
Comment (MartinMao2023) [reply qdax/core/emitters/dcrl_emitter.py:302]: I have updated that to "subkey1, subkey2 = jax.random.split(key)". Here in this function, the key is not passed to any following content.

File: qdax/core/emitters/qpg_emitter.py (hunk 6 above)
Comment (MartinMao2023) [reply qdax/core/emitters/qpg_emitter.py:214]: Actually, only the off-springs. I have made the same changes in dcrl_emitter as well.

File: qdax/core/emitters/qpg_emitter.py (hunk 5 above)
Comment (MartinMao2023) [reply qdax/core/emitters/qpg_emitter.py:202]: Updated that. But again, I see mixture of use of "GARepertoire", "Repertoire", and "MapElitesRepertoire". In this file "qpg_emitter", "GARepertoire" is used in function "emit", and "Repertoire" is used in function "init" and "state_update". 
I recommend unifying all of that into "Repertoire" or "MapElitesRepertoire".

File: qdax/core/emitters/dcrl_emitter.py (hunk 11 above)
Comment (MartinMao2023) [reply qdax/core/emitters/dcrl_emitter.py:185]: They are completely the same, just a variable that renames "None". I like to use None for simplicity.

File: qdax/core/emitters/dcrl_emitter.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Lookatator) [reply qdax/core/emitters/dcrl_emitter.py]: Agreed! In a future PR, we will remove Repertoire

File: qdax/core/emitters/dcrl_emitter.py (hunk 11 above)
Comment (Lookatator) [reply qdax/core/emitters/dcrl_emitter.py:186]: I know they are the same, I just said `jnp.newaxis` is clearer than `None` for people who are not familiar with this syntax. 

I don't think it's big deal anyway, so let's keep it this way.

**Review by Lookatator — APPROVED** [review]
